### ISSUE TITLE
FTP / SFTP Reader

### ISSUE THREAD

**Issue by kyleobrien91** [portphp/portphp#39]
@ddeboer @Baachi @sagikazarmark - I just want your opinion on a new Reader use case. If you agree, I'll get going on writing it. Most importantly, perhaps you can tell me if this is a use case that can be achieved using one of the current readers that are implemented (but I doubt it :P)

A large part of our workflow is allowing people to import data. By far the biggest use is import via CSV.
The second largest use case is the import of files that reside on an SFTP server somewhere.

The reader would essentially follow the spec:
- connect to the FTP / SFTP using an appropriate, well maintained library.
- read the contents of the directory specified in the connection
- optionally, move those files into another remote 'working' directory (like a timestamp folder)
- read the full file paths relative to the connection base into the $items array
- when moving over files, look at optional decompression
- The rest will follow the reader interface.

Let me know if you think there is enough of a use case to include this. Also, if you have any suggestions on FTP/SFTP libraries, that would be welcome too - I have some idea but I'm not sure if you have any minimum criteria in mind.

**Comment by Baachi** [comment]
I rememeber me, that we have a `Source` component in the early stage of this library. So you can fetch a file from a server and/or decompress it.
But sadly i don't remember why this was removed.

**Comment by kyleobrien91** [comment]
@Baachi - thanks for the confirmation - do you believe there is a valid use case that warrants bringing it in?

@ddeboer @sagikazarmark - thoughts?

On a side note - depending on your feelings on how long portphp will take, and if we decide this is valid, should I do this work for ddeboer/data-import  as well?

**Comment by sagikazarmark** [comment]
To be honest, I am not a huge fan of the idea. One of the basic features of this library is low memory footprint and stability. Adding such layer would make this less true and would make the workflow more complicated. My solution for this problem is using a special transport reader which lazily downloads the file upon the first read. To avoid being redundant about configuration (like file paths) you could use a factory which creates the reader for you.

**Comment by kyleobrien91** [comment]
@sagikazarmark - thanks!

I'm not sure which part you believe is _**too much**_ of an overhead. Is it because I indicated that we would move the entire folder to an optional 'working' directory? In retrospect, that would be unacceptable, I agree. 

If we changed that part of the spec, would you be more in favour?

**Comment by sagikazarmark** [comment]
I think I was not clear. We advertise PortPHP as a data processing workflow, and I think it is a pretty good one. As such, it requires an input and produces an output. This way the workflow is quite stable, usually there is a limited number of failures.

Adding a transport layer makes things more complicated. The workflow becomes less stable, configuration becomes harder, the possibly number of failures dramatically increases. (Not to mention the non-predictable amount of resources necessary).

I also had this use case, and I also wanted to suggest this thing. After walking around the problem, I decided not doing so because of the reasons above. In most cases, I was able to extract download jobs separately. When it was not possible, I did what I wrote above: wrapped the actual reader and lazily downloaded the content upon the first read. However, I had quite a few stability problems with this, so lately I am doing downloads separately which triggers the processing upon a successful download. This allows more flexible error handling.
